### PR TITLE
feat(certificatemanager): allow tagging DnsValidatedCertificate

### LINKED ISSUE #12382: (certificatemanager): DnsValidatedCertificate does not allow you to customize tags on the requested certificate
Add a `certificateTags` parameter to `DnsValidatedCertificateProps` which will configure the Lambda to create tags on the requested certificate.

### Use Case

We use permission boundaries to control what CDK can create roles for. As DnsValidatedCertificate creates a role for the function that manages the certificate, we need to be able to limit this role using tags to just the managed certificate.

### Proposed Solution

Include an `acm.addTagsToCertificate` call in the `requestCertificate` function with tags defined in `DnsValidatedCertificateProps`.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by timothy-farestad**
Closes #12382 

Attempting to implement the fix suggested in #12382 to allow the DnsValidatedCertificate resource to be taggable.  Currently, only the custom lambda that is created is tagged, but the certificate provisioned by the lambda is not tagged.  This would allow the lambda to pass tags through to the certificate, too.  

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js [hunk 1]
```diff
@@ -97,6 +98,17 @@ const requestCertificate = async function(requestId, domainName, subjectAlternat
 
   console.log(`Certificate ARN: ${reqCertResponse.CertificateArn}`);
 
+
+  if (!!tags) {
+    const map = new Map(Object.entries(tags));
+    const result = Array.from(map).map(([Key, Value]) => ({ Key, Value }))
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/lambda-packages/dns_validated_certificate_handler/lib/index.js:104]: I believe the `new Map` here is redundant, unless there's a specific (edge) case I'm missing. The below should work:

```suggestion
    const result = Array.from(Object.entries(tags)).map(([Key, Value]) => ({ Key, Value }))
```

File: packages/@aws-cdk/aws-certificatemanager/README.md [hunk 2]
```diff
@@ -34,7 +34,7 @@ See https://docs.aws.amazon.com/acm/latest/userguide/acm-limits.html for more in
 
 DNS validation is the preferred method to validate domain ownership, as it has a number of advantages over email validation.
 See also [Validate with DNS](https://docs.aws.amazon.com/acm/latest/userguide/gs-acm-validate-dns.html)
-in the AWS Certificate Manager User Guide.
+in the AWS Certificate Manager User Guide.  All stack tags will be applied to the DNS validated Certificate.
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/README.md:37]: I think this is a little out of place here, and doesn't really need to be here. Our "all features must have README updates" rule is sometimes a bit aggressive.
```suggestion
in the AWS Certificate Manager User Guide.
```

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts [hunk 3]
```diff
@@ -55,8 +57,15 @@ export interface DnsValidatedCertificateProps extends CertificateProps {
...
  * @experimental
  */
-export class DnsValidatedCertificate extends cdk.Resource implements ICertificate {
+export class DnsValidatedCertificate extends cdk.Resource implements ICertificate, cdk.ITaggable {
   public readonly certificateArn: string;
+
+  /**
+ * Resource Tags.
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-certificatemanager-certificate.html#cfn-certificatemanager-certificate-tags
+ */
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts:66]:
```suggestion
  /**
   * Resource Tags.
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-certificatemanager-certificate.html#cfn-certificatemanager-certificate-tags
   */
```

File: packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts [hunk 4]
```diff
@@ -46,6 +46,8 @@ export interface DnsValidatedCertificateProps extends CertificateProps {
    * @default - A new role will be created
    */
   readonly customResourceRole?: iam.IRole;
+
+
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/lib/dns-validated-certificate.ts:50]:
```suggestion
```

**Review by njlynch — DISMISSED** [review]
Looks great, thanks!

I left a few minor comments. Additionally, could you add two more tests:
1. A test for `handler.test.js` that shows success if tags are not provided. 
2. A test for `dns-validated-certificate.test.ts` showing that tags are included as properties of the CustomResource if they are provided.

**Comment by timothy-farestad** [comment]
Really appreciate the review, @njlynch, I believe this latest commit addresses the tests as well as the syntax changes suggested.

File: packages/@aws-cdk/aws-certificatemanager/test/dns-validated-certificate.test.ts [hunk 5]
```diff
@@ -137,6 +137,40 @@ test('test root certificate', () => {
   });
 });
 
+test('test tags are passed to customresource', () => {
+  const stack = new Stack();
+  let tags = [
+    { key: 'Key1', value: 'Value1' },
+  ];
+  tags.map(({ key, value }) => Tags.of(stack).add(key, value));
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/test/dns-validated-certificate.test.ts:145]: Keep it simple. :)

```suggestion
  Tags.of(stack).add('Key1', 'Value1');
```

File: packages/@aws-cdk/aws-certificatemanager/test/dns-validated-certificate.test.ts [hunk 6]
```diff
@@ -137,6 +137,40 @@ test('test root certificate', () => {
...
+      ],
+    },
+    DomainName: 'example.com',
+    HostedZoneId: {
+      Ref: 'ExampleDotCom4D1B83AA',
+    },
+    Tags:
+      {
+        Key1: 'Value1',
+      },
```
Comment (njlynch) [inline packages/@aws-cdk/aws-certificatemanager/test/dns-validated-certificate.test.ts:170]:
```suggestion
    Tags: {
      Key1: 'Value1',
    },
```

**Review by njlynch — APPROVED** [review]
I made two other minor changes, and this is good to go!

Thanks for the contribution!

**Comment by rix0rrr** [comment]
@mergifyio refresh

**Pull request merged**